### PR TITLE
mars viewer

### PR DESCRIPTION
**Pull request by ilkovecVladislav**

### CODE REVIEW

File: homeworks/07-mars-viewer/public/manifest.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+  "short_name": "React App",
```
Comment (avilde) [inline homeworks/07-mars-viewer/public/manifest.json]: If manifest is used then would be beneficial to also update title here.

File: homeworks/07-mars-viewer/src/reducers/photos.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      api_key: process.env.REACT_APP_API_KEY,
+    }).toString();
+    const { photos } = await fetch(
+      `${'https://api.nasa.gov/mars-photos/api/v1/rovers/perseverance/photos'}?${params}`,
+    ).then((response) => response.json());
+
+    return Promise.resolve({ sol, photos });
+  }
+
+  return Promise.resolve({ sol, photos: [] });
```
Comment (avilde) [inline homeworks/07-mars-viewer/src/reducers/photos.tsx]: If API_KEY is not provided there should be a `Promise.reject` with error message for user to understand what is missing.

File: homeworks/07-mars-viewer/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+[Demo](https://musing-shaw-197471.netlify.app/).
+
```
Comment (avilde) [inline homeworks/07-mars-viewer/README.md:2]: Read me file is missing environment variables setup

`env.example` is not correct environment file it should be `.env.development`.
Also regarding API_KEY, reviewer should not need to ask for it privately via Slack.

File: homeworks/07-mars-viewer/src/App/App.styled.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import styled from '@emotion/styled';
+
+export const Navigation = styled.ul`
+  display: flex;
+  justify-content: center;
+  margin: 0;
+  padding: 0;
+  list-style-type: none;
+  padding-top: 20px;
```
Comment (avilde) [inline homeworks/07-mars-viewer/src/App/App.styled.tsx:9]: For better scaling across different device sizes it is prefered to use `rem` or `em` units for paddings and margins

File: homeworks/07-mars-viewer/src/App/PhotosTab/Form.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    event.preventDefault();
+
+    dispatch(loadPhotos(sol));
+  };
+
+  return (
+    <FormContainer>
+      <FormTitle>Select Sol and press &ldquo;load&rdquo;</FormTitle>
+      <form onSubmit={handleSubmit}>
+        <Input type="number" value={sol} onChange={handleChange} step={1} />
```
Comment (avilde) [inline homeworks/07-mars-viewer/src/App/PhotosTab/Form.tsx]: The minimum value should be `0` as it also returns data.

File: homeworks/07-mars-viewer/src/App/PhotosTab/PhotosGallery.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      if (isFavourite) {
+        dispatch(addToFavourite(id));
+      } else {
+        dispatch(removeFromFavourite(id));
+      }
+    },
+    [dispatch],
+  );
+
+  if (Array.isArray(photos)) {
```
Comment (avilde) [inline homeworks/07-mars-viewer/src/App/PhotosTab/PhotosGallery.tsx]: Also `length > 0` should be checked. What will show on UI if `photos.length === 0` ?

File: homeworks/07-mars-viewer/src/assets/icons/unicorn.svg [hunk 7]
```diff
@@ -0,0 +1 @@
...-.192-4.151-.068l-8.199-25.906c4.819-.882 9.88-1.467 15.029-1.688 28.001-1.207 52.484 22.757 54.876 44.164.613 5.489 5.568 9.44 11.049 8.828 5.488-.613 9.441-5.559 8.828-11.049-1.755-15.704-10.814-31.584-24.854-43.571zm-69.377 32.194c-20.247 5.978-27.3 5.544-30.188-1.859-1.586-4.064-1.33-7.354.83-10.665 3.525-5.404 11.685-10.185 22.127-13.481l8.145 25.737c-.306.089-.598.174-.914.268z" fill="#3c122c"/><path d="m84.274 239.566c-2.832 0-5.352-2.014-5.891-4.898-.609-3.258 1.538-6.392 4.795-7l40.838-7.633c3.268-.615 6.392 1.538 7 4.795.609 3.258-1.538 6.392-4.795 7l-40.838 7.633c-.373.07-.744.103-1.109.103z" fill="#fff"/><path d="m177.4 277.542c-2.832 0-5.352-2.014-5.891-4.898-.609-3.258 1.538-6.392 4.795-7l40.838-7.633c3.269-.612 6.392 1.538 7 4.795.609 3.258-1.538 6.392-4.795 7l-40.838 7.633c-.373.07-.744.103-1.109.103z" fill="#fff"/><path d="m151.13 311.613c-2.082 0-4.104-1.084-5.213-3.02-1.159-2.025-3.281-3.26-5.655-3.196-1.944.046-3.735.963-4.912 2.517-2 2.643-5.765 3.162-8.405 1.161-2.642-2.001-3.162-5.764-1.161-8.405 3.4-4.489 8.575-7.139 14.201-7.269 6.721-.191 12.997 3.381 16.347 9.229 1.646 2.876.65 6.542-2.226 8.189-.94.538-1.965.794-2.976.794z" fill="#3c122c"/></g></g></svg>
```
Comment (avilde) [inline homeworks/07-mars-viewer/src/assets/icons/unicorn.svg]: To be honest, the unicorn image is super distracting and due to nature of lazy-load the SVG is visible only for split second.

File: homeworks/07-mars-viewer/src/reducers/favourite.tsx [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+interface FavouriteState {
+  list: number[];
+}
+
+const initialState: FavouriteState = {
+  list: [],
+};
+
+export const counterSlice = createSlice({
```
Comment (avilde) [inline homeworks/07-mars-viewer/src/reducers/favourite.tsx]: The name of slice is wrong, is it a leftover for using previous homework as template?

File: homeworks/07-mars-viewer/src/reducers/selectors.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    return photos.length > 0 ? photos : 'No such photos found';
+  },
+);
+
+export const useIsLoading = (): boolean =>
+  useSelector((state: RootState) => state.photos.isLoading);
+export const useSelectedSol = (): string => useSelector(selectedSolSelector);
+export const useFavouritesIds = (): number[] => useSelector(favouritesSelector);
+export const useCurrentSolPhotos = (): string | Photo[] => useSelector(currentSolPhotosSelector);
+export const useFavouritesPhotos = (): string | Photo[] => useSelector(favouritesPhotosSelector);
```
Comment (avilde) [inline homeworks/07-mars-viewer/src/reducers/selectors.ts]: The naming convention could be improved - two plural worlds is more difficult to read.
`useFavouritePhotos` would sound better.

File: homeworks/07-mars-viewer/src/types/Photo.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+type Photo = {
```
Comment (avilde) [inline homeworks/07-mars-viewer/src/types/Photo.ts]: I would strongly suggest writing a conversion util function which takes `PhotoJson[]` data and converts to `Photo[]` data and use only the necessary properties + rename them for more front-end oriented naming conventions like `camelCase`.
You are only reading the data from NASA db and not writing anything.

**Review by avilde — APPROVED** [review]
## Review

# Good parts
* clean code style 💯 
* application works (NASA data can be retrieved and is displayed quite well)
* use of lazyload attribute for images 👍🏻 
* use of styled components

# What can be improved
* the unicorn SVG should be removed as it doesn't provide any benefit to user, only distracting and confusing.
* the `store` files should be structures better. Usually the composition is following:
```
<src>\store
  |_ /actions
  |_ /reducers
  |_ /middleware
  |_ store.ts (or index.ts)
```

# Summary
* The application is well done. Good job, Vladislav!

File: homeworks/07-mars-viewer/src/reducers/selectors.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    const selectedSolsPhotos: Photo[] | undefined = allSols[selectedSol];
+
+    if (selectedSolsPhotos && selectedSolsPhotos.length > 0) {
+      return selectedSolsPhotos;
+    }
+    if (selectedSolsPhotos && selectedSolsPhotos.length === 0) {
+      return 'No photos for this sol :(';
+    }
+
+    return 'Photos are not loaded';
```
Comment (ivankoleda) [inline homeworks/07-mars-viewer/src/reducers/selectors.ts]: I would add that it's better to move returning strings out of this selector, it real world app one selector could be used multiple times in different components, and this result may not be suitable for other use cases. Also it's better to return only one type of data array in this case

File: homeworks/07-mars-viewer/src/reducers/selectors.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+    const photos = allPhotos.filter((photo) => favouritesIds.includes(photo.id));
+
+    return photos.length > 0 ? photos : 'No such photos found';
+  },
+);
+
+export const useIsLoading = (): boolean =>
+  useSelector((state: RootState) => state.photos.isLoading);
+export const useSelectedSol = (): string => useSelector(selectedSolSelector);
```
Comment (ivankoleda) [inline homeworks/07-mars-viewer/src/reducers/selectors.ts]: this custom hooks are redundant, reselect is able to infer return types and `useSelector(selectedSolSelector)` is easy understandable by any react dev that something is retrieved from redux, `useSelectedSol` may be getting data from state, from context, from redux, or any other state

**Pull request merged**